compiler/nimtypes.nim

```diff
@@ -0,0 +1,480 @@
+  # consider renaming as `tyAbstractVarRange`
+  abstractVarRange* = {tyGenericInst, tyRange, tyVar, tyDistinct, tyOrdinal,
+                       tyTypeDesc, tyAlias, tyInferred, tySink, tyOwned}
+  abstractInst* = {tyGenericInst, tyDistinct, tyOrdinal, tyTypeDesc, tyAlias,
+                   tyInferred, tySink, tyOwned} # xxx what about tyStatic?
+
+const
+  TypeKindBits = 8'u32
+  TypeKindMask = (1'u32 shl TypeKindBits) - 1'u32
+
+type
+  TypeNode* = object     # 4 bytes
+    x: uint32
+
+template kind*(n: TypeNode): TTypeKind = TTypeKind(n.x and TypeKindMask)
+template operand(n: TypeNode): uint32 = (n.x shr TypeKindBits)
+
+proc integralBits*(n: TypeNode): int {.inline.} =
+  # Number of bits in the IntTy, etc. Only valid for integral types.
+  assert n.kind in {tyInt, tyInt8, tyInt16, tyInt32, tyInt64,
+    tyFloat, tyFloat32, tyFloat64, tyFloat128,
+    tyUInt, tyUInt8, tyUInt16, tyUInt32, tyUInt64}
+  result = int(n.operand)
+
+template toX(k: TTypeKind; operand: uint32): uint32 =
+  uint32(k) or (operand shl TypeKindBits)
+
+template toX(k: TTypeKind; operand: LitId): uint32 =
+  uint32(k) or (operand.uint32 shl TypeKindBits)
+
+type
+  TTypeAttachedOp* = enum ## as usual, order is important here
+    attachedWasMoved,
+    attachedDestructor,
+    attachedAsgn,
+    attachedDup,
+    attachedSink,
+    attachedTrace,
+    attachedDeepCopy
+
+  TType* = object             # Keep in sync with PackedType
+    owner*: PSym              # the 'owner' of the type
+    sym*: PSym                # types have the sym associated with them
+                              # it is used for converting types to strings
+    size*: BiggestInt         # the size of the type in bytes
+                              # -1 means that the size is unkwown
+    align*: int16             # the type's alignment requirements
+    paddingAtEnd*: int16      #
+    loc*: TLoc
+    typeInst*: PType          # for generic instantiations the tyGenericInst that led to this
+                              # type.
+
+type
+  Literals* = object
+    strings*: BiTable[string]
+    numbers*: BiTable[int64]
+
+  TypeGraph* = object
+    nodes: seq[TypeNode]
+    lit: Literals
+    ext*: Table[PType, TType]
+    flags*: Table[PType, TTypeFlags]
+    n*: Table[PType, PNode]
+    callConv*: Table[PType, TCallingConvention]
+
+proc prepareExt*(g: var TypeGraph; t: PType): var TType =
+  result = g.ext.mgetOrPut(t, default(TType))
+
+const
+  VoidId* = PType 0
+  Bool8Id* = PType 1
+  Char8Id* = PType 2
+  LastBuiltinId* = 3
+
+proc initTypeGraph*(lit: Literals): TypeGraph =
+  result = TypeGraph(nodes: @[
+    TypeNode(x: toX(tyVoid, 0'u32)),
+    TypeNode(x: toX(tyBool, 8'u32)),
+    TypeNode(x: toX(tyChar, 8'u32)),
+  ], lit: lit)
+  assert result.nodes.len == LastBuiltinId
+
+type
+  TypePatchPos* = distinct int
+
+const
+  InvalidTypePatchPos* = TypePatchPos(-1)
+
+  TypeAtoms = {tyNone, tyBool, tyChar,
+    tyEmpty, tyNil,
+    tyInt, tyInt8, tyInt16, tyInt32, tyInt64,
```

Comment:
time for parametrized int? so as to prep for a future with custom bit lengths.. and a vector type for simd where int is a special case of a length-1 simd type

Reply:
It is parametrized indeed, but well hidden so far. ;-)